### PR TITLE
Update JSON serialization code

### LINKED ISSUE #19575: Reinforced items, damage error?
I'm not sure if this is normal as far as the game goes, or is genuinely a bug, so someone clarify if possible. 

Updated from version... I think #5975 or #5976 (according to http://ci.narc.ro/view/Cataclysm-DDA/job/Cataclysm-Matrix/changes ) to the most recent version. Caused a bunch of items to vanish.

All the items in question were items that had been reinforced via tailoring kit or welder, and had the line 

"damage" : -1.000000e+000," 

in the save file.

### PR DESCRIPTION
**Pull request by mugling**
We don't leverage `stringstream` sufficiently and it looks like a lot of our custom parsing is very brittle. 

Fixes #19575

### CODE REVIEW

**Comment by mugling** [comment]
What is our lowest targeted version of `gcc`? I'd like to use `std::quoted` but that is C++14. It's supported by MSCV14 (2015) but only by gcc 5

**Comment by illi-kun** [comment]
```doc/COMPILER_SUPPORT.md``` states that

> GCC is the preferred compiler for building on Linux. In order to fully support the C++11 standard, we are using GCC version 4.8.2.

**Comment by Coolthulhu** [comment]
As far as I recall, 4.9.

**Comment by mugling** [comment]
> As far as I recall, 4.9.

That looks like the version that Code::Blocks ships with

**Comment by mugling** [comment]
Ready. A lot simpler than the previous implementation and with better documentation

File: src/json.cpp [hunk 1]
```diff
@@ -1563,30 +1563,21 @@ std::string JsonIn::substr(size_t pos, size_t len)
...
+    stream->setf( std::ios_base::boolalpha );
 }
 
 void JsonOut::write_indent()
 {
-    const char indent[5] = "    ";
-    for (int i = 0; i < indent_level; ++i) {
-        stream->write(indent, 4);
-    }
+    std::fill_n( std::ostream_iterator<char>( *stream ), indent_level * 2, ' ' );
```
Comment (Coolthulhu) [inline src/json.cpp]:
```
'ostream_iterator' is not a member of 'std'
```

Comment (mugling) [inline src/json.cpp]: Probably a missing include, try `#include <iterator>`

Comment (mugling) [inline src/json.cpp]: Have pushed that in 476fc88

**Pull request merged**